### DISCUSSION TITLE
is there a way to get segment value in C code?

### DISCUSSION THREAD

**Discussion by LowLevelMahn** [open-watcom/open-watcom-v2#1147] [General]
i try to replace a reverse engineered data-segment (old DOS game exe) with a 100% equal C equivalent

and i've got variables that contain segment,offset values like in this simple example

```
.model medium

my_data_seg segment 'DATA'

print0_msg db 'print0!',13,10,'$'

test_seg1 dw seg my_code_seg0   ; <====== ...
test_seg2 dw seg my_code_seg1
test_seg3 dw seg my_data_seg
test_seg4 dw seg my_stack_seg
test_ofs1 dw offset print0_msg

my_data_seg ends

my_code_seg0 segment 'CODE'
; ...
my_code_seg0 ends

my_code_seg1 segment 'CODE'
; ...
my_code_seg1 ends

my_stack_seg segment stack 'STACK'
  db 128 dup (?)
my_stack_seg ends

end start
```

are there some extensions, macros in Watcom C available to get the same values in C variables - when im combining assembler obj and C obj files?

im aware of __segname and __segment, __based and that stuff but i don't know how to combine them to get the result

Turbo C got something like that (http://bitsavers.informatik.uni-stuttgart.de/pdf/borland/turbo_c/Turbo_C_Users_Guide_1987.pdf page 249)

```
There are also four special near data pointers: _cs, _ds, _ss, and _es.
These are 16-bit pointers that are specifically associated with the
corresponding segment register. For example, if you were to declare a
pointer to be
char _ss *Pi 
then p would contain a 16-bit offset into the stack segment.
```

so something like

`char segmentof("my_code_seg0") *test_seg1;`

**Comment by jmalak** [comment]
If you search equivalent of Turbo C predefined segments then it is in OW described here
[default segments](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#Segment_Constant_Based_Pointers_and_Objects)
```
     "_CODE"
     "_CONST"
     "_DATA"
     "_STACK"
```

it is not pointers it is segment names that you need translate to segment value.
Operator __segname( .... ) do this. The same you can do for your own segment names.
```
__segname( "segment name")
```

Next you need to combine segment value with offset to pointer.
You can use _MK_FP macro or based pointer it depends of what form you want use.
Based pointers are near pointer in memory (hold only offset value) and compiler know for which segment it is.
On access such pointer variable compiler construct real far pointer from known based segment value (fix value or variable value) and based pointer value (offset part). But compiler need everytime add code to calculate far pointer on run-time if you access memory to which it points. If you don't need to save memory then far pointers is best solution (no code overhead).

Example for pointer to "MySEG" segment by both methods (including initialization).
```
char __based( __segname( "MySEG" ) ) *thing = 0x100;
char __far * thing = _MK_FP( __segname( "MySEG" ), 0x100 );
```

If you have variable which hold the segment value then the construct should looks like
```
__segment my_seg;

char __based( my_seg ) *thing = 0x100;
char __far * thing = _MK_FP( my_seg, 0x100 );
```

But if you need to define variables to be in some segment outside default _DATA segment then you can use construct like.
```
char __based( __segname( "MySEG" ) ) thing;
```

It add variable "thing" into segment "MySEG".

Take into account that only default "_DATA" data segment is near and all your other data segments are far therefore you need to use far pointers to access variables in these segments ( based pointers or true far pointers ).
Of cause you can use this construct only with segment names because compiler need know segment name where it will be placed(allocated).

there are other possibility to manipulate with default segment names, see compiler -n.. options in OW documentation
[-nc](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#ncEXnameY)
[-nd](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#ndEXnameY)
[-nm](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#nmEXnameY)
[-nt](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#ntEXnameY)

There are also appropriate #pragma variants which can be used selectively (see OW documentation).

**Reply by LowLevelMahn** [reply]
thanks - that works for me
